### ISSUE TITLE
Naming the cards

### ISSUE THREAD

**Issue by dnaidionov** [cardstack/cardstack#986]
_**Update**: accommodated devs' comments_'

This ticket describes work that should be done in the Hub.

To-do to close this ticket:
- [x] add id, card-name, and route as properties that exist on all cards
- [x] assigns the id to the route, as a stopgap
- [ ] make card name user-editable (does not matter if this is unique)
- [ ] provide a default value for card name
- [ ] make sure that the id is server-side generated, unique, and not editable
- [ ] open a ticket that describes what you think should be done next, including UI
- [ ] use the card name in the card header (i.e. in Schema mode, the bright blue card header should say "My Card Name" not "My-Card-Name"

# General card properties

All cards should have the following properties: 

- id 
- card-name
- route 

## `id`
- id is generated by the system
- it has to be unique 
- id cannot be changed
- id is internal and is not shown to the user

## `card-name`
This is the name that users give to the card (the instance of the card) they create - whether it's a regular instance of a card or a template. The name that users give to the card becomes the type for all cards that adopt this card. 

- card-name doesn't have to be unique
- card-name is required
- card-name can be changed

## `route`
- it has to be unique
- by default, it would be a combination of dasherized `card-name` and _(optional)_ some modifier to make it unique - populated as soon as user enters card name (or even in sync with it). 
E.g. if user enters "invoice template" into the card name field, and there is already a card with route "invoice-template", we will suggest modified route "invoice-template-284" that user can still edit
- it can be changed by users provided it maintains uniqueness

The exact method of ensuring route uniqueness is TBD.  
As user types the route we shall validate its uniqueness (whether as user types, with debounce, 
or on blur) and, if validation failed, suggest a modified version that is unique.
For users' convenience, we will label this field "Card URL" since non-developers might not be familiar with routing concept. 

## aliases/vanity URLs
In the future, we might create vanity URLs for cards, but that doesn't change the card route. 

- vanity URL has to be unique (same requirement as for `route`)

# Designs

## Regular 
![image](https://user-images.githubusercontent.com/1836666/68807300-991bf100-0635-11ea-946a-9dca417b9c62.png)

## Route validation failed
![image](https://user-images.githubusercontent.com/1836666/68807314-9f11d200-0635-11ea-842c-5bd55f3117b3.png)
Error state and message styling are in draft. Error state design will be addressed in a separate ticket. 

XD file: https://drive.google.com/open?id=1QSAC1931gocLR_sgbHqPQIAiQ9oegQ5C

**Comment by habdelra** [comment]
an ID cannot be changed--full stop. it's a primary key to a card.

**Comment by habdelra** [comment]
Also the convention is that we dasherize our fields. so instead of `cardName` above, `card-name` would be the actual field name.

Also, yes, within a card, amongst both its own fields and fields that it adopts, all fields are unique. a card ID is not a field. but you could have a field called `id` if you the card author wants (albiet, it would be a really bad name for a field).

**Comment by habdelra** [comment]
Also we need to consider how we enforce the uniqueness constraint on ID's and routes (aka slugs) for cards. If a user is creating a card's ID then we need to make sure it doesn't collide with an existing ID. same for a route, a user cannot set a route for a card to be the same as another card is using.

We could decide that ID's are auto generated UUID's (or something similar) and that alleviates the need for a user to worry about specifying duplicate ID's.

**Comment by bagby** [comment]
I agree with everything @habdelra said, and would go a step further and say ID's should *never* be created by a user (and they should *never* be modified once created).

**Comment by bagby** [comment]
@dnaidionov I believe this is blocked at the moment by not being able to search for cards by field, for example to ensure uniqueness of `route` we would need cards to be searchable by route. Can you confirm/deny @habdelra ?

**Comment by habdelra** [comment]
confirm

**Comment by jenweber** [comment]
Blocked by search feature.

**Comment by jenweber** [comment]
I also need a way to generate the IDs. Two options I can think of:
1. rely on the popular `uuid` package https://www.npmjs.com/package/uuid
2. Use native crypto methods ([Example](https://stackoverflow.com/a/40191779/8585664)) ` crypto.randomBytes(16).toString("hex")`

The native randombytes method is used in the uuid package. Since we don't need these to be anything fancy, just reasonably-likely-to-be-unique IDs, I think I will use the native crypto method directly, which we use elsewhere in the mono repo. If anyone else prefers something else, let me know.

**Comment by jenweber** [comment]
Also, since route uniqueness is blocked for now, I will just default to using the id, and work to implement the other aspects.

**Comment by jenweber** [comment]
So, to do any meaningful work on this issue, we will need to rethink how we do tests, _or_ implement the route feature that is blocked by search. The reason is that every test relies heavily on declaring the id that should be used (e.g. Millennial Puppies), but this PR includes generating server-side ids. We could shift this to use the route, when it is available. I could PR adding the card name and route attributes to the card constructor, but that would be basically cruft until we add at least id generation or routes.

I have some work on my machine locally that I don't want to push because CI will break. For my own future reference, the branch is `naming-cards-986`.

**Comment by jenweber** [comment]
As discussed in today's standup, we should get search & routing before we pick this up again.
